### PR TITLE
71-expose-project-management-though-cli

### PR DESCRIPTION
**Pull request by Pbarbosa92**
This PR adds the commands to list and download files from projects to the CLI.

## How to use
List all projects with `inductiva projects list`

Download all files from every task in a project `inductiva projects download project-name`
Download all files from every task in a project to downlods dir `inductiva projects download project-name --output-dir downloads`
Download file1.txt from every task in a project `inductiva projects download project-name --files file1.txt`
Download stdout.txt and stderr.txt from every task in a project `inductiva projects download project-name --std`

### CODE REVIEW

File: inductiva/_cli/cmd_projects/download.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    This function downloads the files of the tasks of a project. The project
+    name is specified with the --project-name flag. The files to download can
+    be specified with the --files flag. If no files are specified, all files are
+    downloaded. The standard output and error files can be downloaded with the
+    --std flag. The files are saved in the default directory (inductiva-output)
+    if no output directory is specified with the --output-dir flag.
+
+    The downloads are done in parallel using a ThreadPoolExecutor with a maximum
+    of 100 workers.
```
[the remark this answers is not in the document]
Comment (Pbarbosa92) [reply inductiva/_cli/cmd_projects/download.py:19]: Reduced to 10.

**Comment by Pbarbosa92** [comment]
Closes https://github.com/inductiva/tasks/issues/71

File: inductiva/_cli/cmd_projects/download.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    of 10 workers.
+    """
+
+    project_name = args.project_name
+    output_dir = args.output_dir
+    files = args.files
+    std = args.std
+
+    if std:
+        files = ["stdout.txt", "stderr.txt"]
```
Comment (Franreno) [inline inductiva/_cli/cmd_projects/download.py:28]: Are the --std and --files exclusive? 

The std will overwrite the files if both arguments are used.. should we inform this to the user?

File: inductiva/_cli/cmd_projects/download.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    project_tasks = project.get_tasks()
+
+    if output_dir is not None:
+        inductiva.set_output_dir(output_dir)
+
+    futures = []
+
+    with cf.ThreadPoolExecutor(max_workers=10) as executor:
+        for task in project_tasks:
+            future = executor.submit(task.download_outputs, filenames=files)
```
Comment (Franreno) [inline inductiva/_cli/cmd_projects/download.py:40]: This makes a lot of text into the terminal
Screenshot 2024-06-18 at 09 50 36

Should we wrap this ThreadPool into a tqdm?

File: inductiva/_cli/cmd_projects/list.py [hunk 4]
```diff
@@ -10,14 +10,14 @@
 
 
 def get_projects(_, fout: TextIO = sys.stdout):
-    """ Lists the user's projetcs. 
+    """ Lists the user's projects. 
 
-    Lists all the user's projetcs.
+    Lists all the user's projects.
```
Comment (Franreno) [inline inductiva/_cli/cmd_projects/list.py:15]: These comments seems to be repetitive, maybe we can choose one of them

**Review by Franreno — COMMENTED** [review]
Left some comments regarding one doubt with the usage of `--files` and `--std`

File: inductiva/_cli/cmd_projects/download.py (hunk 2 above)
Comment (Pbarbosa92) [reply inductiva/_cli/cmd_projects/download.py:28]: Yes, you cant use both of them.

File: inductiva/_cli/cmd_projects/download.py (hunk 3 above)
Comment (Pbarbosa92) [reply inductiva/_cli/cmd_projects/download.py:40]: This behaviour is defined in the `task.download_output`
We have a progress bar if we download all files.

**Review by Franreno — APPROVED** [review]

**Comment by Pbarbosa92** [comment]
@Franreno Can you do some QA ?
Test the following commands
inductiva projects list
inductiva projects download project-name
inductiva projects download project-name --output-dir downloads
inductiva projects download project-name --files file1.txt
inductiva projects download project-name --std

**Comment by bgregoinductiva** [comment]
I was already in the process of doing some QA.

The command `inductiva projects download test_project --output-dir downloads` is not working correctly. It downloads the output to the default directory instead of the specified one.

Everything else is working correctly. I confirmed the downloads are being done by inspecting the output directory.

```
(env_tests) brunogrego@oxygen:~/inductiva/test_scripts$ inductiva projects list

 NAME             NR_TASKS
 test_project     2
 bgrego49a0d2f3   584

(env_tests) brunogrego@oxygen:~/inductiva/test_scripts$ inductiva projects download test_project
Downloading simulation outputs to inductiva_output/ndenabviysxzdc9vj37kc9q4t/output.zip...
Downloading simulation outputs to inductiva_output/atb5h4pztgyobcd3frwsn70b9/output.zip...
100%|████████████████████████████████████████████████████████████████| 879k/879k [00:00<00:00, 6.01MB/s]
Uncompressing the outputs to inductiva_output/ndenabviysxzdc9vj37kc9q4t...
100%|████████████████████████████████████████████████████████████████| 878k/878k [00:00<00:00, 5.81MB/s]
Uncompressing the outputs to inductiva_output/atb5h4pztgyobcd3frwsn70b9...
Downloads completed.

(env_tests) brunogrego@oxygen:~/inductiva/test_scripts$ inductiva projects download test_project --output-dir downloads
/home/brunogrego/inductiva/env_tests/lib/python3.10/site-packages/inductiva/tasks/task.py:707: UserWarning: Path already exists, files may be overwritten.
  warnings.warn("Path already exists, files may be overwritten.")
Downloading simulation outputs to inductiva_output/ndenabviysxzdc9vj37kc9q4t/output.zip...
Downloading simulation outputs to inductiva_output/atb5h4pztgyobcd3frwsn70b9/output.zip...
100%|████████████████████████████████████████████████████████████████| 879k/879k [00:00<00:00, 6.93MB/s]
Uncompressing the outputs to inductiva_output/ndenabviysxzdc9vj37kc9q4t...   | 0.00/878k [00:00<?, ?B/s]
100%|████████████████████████████████████████████████████████████████| 878k/878k [00:00<00:00, 5.83MB/s]
Uncompressing the outputs to inductiva_output/atb5h4pztgyobcd3frwsn70b9...k/878k [00:00<00:00, 2.03MB/s]
Downloads completed.

(env_tests) brunogrego@oxygen:~/inductiva/test_scripts$ inductiva projects download test_project --files conf.gro
/home/brunogrego/inductiva/env_tests/lib/python3.10/site-packages/inductiva/tasks/task.py:707: UserWarning: Path already exists, files may be overwritten.
  warnings.warn("Path already exists, files may be overwritten.")
Downloading simulation outputs to inductiva_output/ndenabviysxzdc9vj37kc9q4t...
Downloading simulation outputs to inductiva_output/atb5h4pztgyobcd3frwsn70b9...
Partial download completed to inductiva_output/ndenabviysxzdc9vj37kc9q4t.
Partial download completed to inductiva_output/atb5h4pztgyobcd3frwsn70b9.
Downloads completed.

(env_tests) brunogrego@oxygen:~/inductiva/test_scripts$ inductiva projects download test_project --std
/home/brunogrego/inductiva/env_tests/lib/python3.10/site-packages/inductiva/tasks/task.py:707: UserWarning: Path already exists, files may be overwritten.
  warnings.warn("Path already exists, files may be overwritten.")
Downloading simulation outputs to inductiva_output/ndenabviysxzdc9vj37kc9q4t...
Downloading simulation outputs to inductiva_output/atb5h4pztgyobcd3frwsn70b9...
Partial download completed to inductiva_output/ndenabviysxzdc9vj37kc9q4t.
Partial download completed to inductiva_output/atb5h4pztgyobcd3frwsn70b9.
Downloads completed.
```

**Comment by Franreno** [comment]
I have a minor problem.

I deleted all of my storage before trying to download everything, since i had over 1k tasks. Then I ran 10 reef3d tasks. This resulted on the command `inductiva storage list -m 1000` just outputing 10 tasks, which is correct. 

My project however still maintains the correct amount of total tasks i've ran:

```
➜  inductiva projects list                   
 NAME                NR_TASKS
 fnogueira5ef4f23f   1225
```

When I use the download command, it only downloads the 10 tasks that I have in my storage. This behavior is normal I think. I would just suggest a new column on the projects list like `NR_TASKS_IN_STORAGE`. In this case it would be something like

```
➜  inductiva projects list                   
 NAME                NR_TASKS     NR_TASKS_IN_STORAGE
 fnogueira5ef4f23f   1225       10
```

Screenshot 2024-06-18 at 13 02 29


---
### Normal download

Other than that, the 10 tasks I had in storage were download successfully.

---

### Using std

Using --std flag worked correctly

---

### Specific files

Download only specific files from all of the tasks also worked.

`inductiva projects download fnogueira5ef4f23f --files REEF3D_Log-Wave/REEF3D-Wave-Timeseries-1.dat`

--- 

### Output dir

Using --output-dir didn't work. It created the inductiva_output folder. Seems like it named the .zip file to the --output-dir parameter.

```
franreno in inductiva/bugbash4/tmp 
➜  ls
output
franreno in inductiva/bugbash4/tmp 
➜  inductiva projects download fnogueira5ef4f23f --output-dir output
Downloading simulation outputs to inductiva_output/uekn3wsqu5w8mnk7j9m6ofuvz/output.zip...
Downloading simulation outputs to inductiva_output/92zpp7nicjsmkig5u122ynu5h/output.zip...
Downloading simulation outputs to inductiva_output/n9h5d2j3jkobyrfnromqycwx4/output.zip...
➜  ls
inductiva_output output
```

---

### Suggestion

Maybe not on the scope of this PR, but I think it would be a good use case to state the amount of tasks to donwload. 
`inductiva projects download fnogueira5ef4f23f` by default downloads all file from the project. Maybe we could specifiy to only download the last N amount of tasks, e.g:
`inductiva projects download fnogueira5ef4f23f -n 10` -> would donwload only the last 10 tasks of this project

**Review by Franreno — CHANGES_REQUESTED** [review]
--output-dir not working properly

**Comment by Franreno** [comment]
### output dir

➜  inductiva projects download fnogueira5ef4f23f --output-dir output
Downloading simulation outputs to output/javb78trqdgjti23gk2myzphv/output.zip...
Downloading simulation outputs to output/92zpp7nicjsmkig5u122ynu5h/output.zip...
Downloading simulation outputs to output/n9h5d2j3jkobyrfnromqycwx4/output.zip...

Working correctly now ✅

**Review by Franreno — APPROVED** [review]

**Comment by Pbarbosa92** [comment]
> I have a minor problem.
> 
> I deleted all of my storage before trying to download everything, since i had over 1k tasks. Then I ran 10 reef3d tasks. This resulted on the command `inductiva storage list -m 1000` just outputing 10 tasks, which is correct.
> 
> My project however still maintains the correct amount of total tasks i've ran:
> 
> ```
> ➜  inductiva projects list                   
>  NAME                NR_TASKS
>  fnogueira5ef4f23f   1225
> ```
> 
> When I use the download command, it only downloads the 10 tasks that I have in my storage. This behavior is normal I think. I would just suggest a new column on the projects list like `NR_TASKS_IN_STORAGE`. In this case it would be something like
> 
> ```
> ➜  inductiva projects list                   
>  NAME                NR_TASKS     NR_TASKS_IN_STORAGE
>  fnogueira5ef4f23f   1225       10
> ```
> 
> Screenshot 2024-06-18 at 13 02 29
> ### Normal download
> Other than that, the 10 tasks I had in storage were download successfully.
> 
> ### Using std
> Using --std flag worked correctly
> 
> ### Specific files
> Download only specific files from all of the tasks also worked.
> 
> `inductiva projects download fnogueira5ef4f23f --files REEF3D_Log-Wave/REEF3D-Wave-Timeseries-1.dat`
> 
> ### Output dir
> Using --output-dir didn't work. It created the inductiva_output folder. Seems like it named the .zip file to the --output-dir parameter.
> 
> ```
> franreno in inductiva/bugbash4/tmp 
> ➜  ls
> output
> franreno in inductiva/bugbash4/tmp 
> ➜  inductiva projects download fnogueira5ef4f23f --output-dir output
> Downloading simulation outputs to inductiva_output/uekn3wsqu5w8mnk7j9m6ofuvz/output.zip...
> Downloading simulation outputs to inductiva_output/92zpp7nicjsmkig5u122ynu5h/output.zip...
> Downloading simulation outputs to inductiva_output/n9h5d2j3jkobyrfnromqycwx4/output.zip...
> ➜  ls
> inductiva_output output
> ```
> 
> ### Suggestion
> Maybe not on the scope of this PR, but I think it would be a good use case to state the amount of tasks to donwload. `inductiva projects download fnogueira5ef4f23f` by default downloads all file from the project. Maybe we could specifiy to only download the last N amount of tasks, e.g: `inductiva projects download fnogueira5ef4f23f -n 10` -> would donwload only the last 10 tasks of this project

This is a problem from managing the storage using storage instead of using the tasks. Using storage to delete folders will not update the database and will create all kinds of issues.

**Comment by bgregoinductiva** [comment]
The command `inductiva projects download test_project --output-dir downloads` now works correctly. But the default command now downloads the outputs to the current directory instead of the default output directory.

```
(env_tests) brunogrego@oxygen:~/inductiva/test_scripts$ inductiva projects download test_project --output-dir downloads
Downloading simulation outputs to downloads/ndenabviysxzdc9vj37kc9q4t/output.zip...
Downloading simulation outputs to downloads/atb5h4pztgyobcd3frwsn70b9/output.zip...
100%|████████████████████████████████████████████████████████████████| 878k/878k [00:00<00:00, 6.04MB/s]
Uncompressing the outputs to downloads/atb5h4pztgyobcd3frwsn70b9...  | 163k/879k [00:00<00:00, 1.61MB/s]
100%|████████████████████████████████████████████████████████████████| 879k/879k [00:00<00:00, 5.17MB/s]
Uncompressing the outputs to downloads/ndenabviysxzdc9vj37kc9q4t...
Downloads completed.

(env_tests) brunogrego@oxygen:~/inductiva/test_scripts$ inductiva projects download test_project
Downloading simulation outputs to /home/brunogrego/inductiva/test_scripts/atb5h4pztgyobcd3frwsn70b9/output.zip...
Downloading simulation outputs to /home/brunogrego/inductiva/test_scripts/ndenabviysxzdc9vj37kc9q4t/output.zip...
100%|████████████████████████████████████████████████████████████████| 878k/878k [00:00<00:00, 5.87MB/s]
Uncompressing the outputs to /home/brunogrego/inductiva/test_scripts/atb5h4pztgyobcd3frwsn70b9..., ?B/s]
100%|████████████████████████████████████████████████████████████████| 879k/879k [00:00<00:00, 5.95MB/s]
Uncompressing the outputs to /home/brunogrego/inductiva/test_scripts/ndenabviysxzdc9vj37kc9q4t...23MB/s]
Downloads completed.
```

**Comment by Pbarbosa92** [comment]
> The command `inductiva projects download test_project --output-dir downloads` now works correctly. But the default command now downloads the outputs to the current directory instead of the default output directory.
> 
> ```
> (env_tests) brunogrego@oxygen:~/inductiva/test_scripts$ inductiva projects download test_project --output-dir downloads
> Downloading simulation outputs to downloads/ndenabviysxzdc9vj37kc9q4t/output.zip...
> Downloading simulation outputs to downloads/atb5h4pztgyobcd3frwsn70b9/output.zip...
> 100%|████████████████████████████████████████████████████████████████| 878k/878k [00:00<00:00, 6.04MB/s]
> Uncompressing the outputs to downloads/atb5h4pztgyobcd3frwsn70b9...  | 163k/879k [00:00<00:00, 1.61MB/s]
> 100%|████████████████████████████████████████████████████████████████| 879k/879k [00:00<00:00, 5.17MB/s]
> Uncompressing the outputs to downloads/ndenabviysxzdc9vj37kc9q4t...
> Downloads completed.
> 
> (env_tests) brunogrego@oxygen:~/inductiva/test_scripts$ inductiva projects download test_project
> Downloading simulation outputs to /home/brunogrego/inductiva/test_scripts/atb5h4pztgyobcd3frwsn70b9/output.zip...
> Downloading simulation outputs to /home/brunogrego/inductiva/test_scripts/ndenabviysxzdc9vj37kc9q4t/output.zip...
> 100%|████████████████████████████████████████████████████████████████| 878k/878k [00:00<00:00, 5.87MB/s]
> Uncompressing the outputs to /home/brunogrego/inductiva/test_scripts/atb5h4pztgyobcd3frwsn70b9..., ?B/s]
> 100%|████████████████████████████████████████████████████████████████| 879k/879k [00:00<00:00, 5.95MB/s]
> Uncompressing the outputs to /home/brunogrego/inductiva/test_scripts/ndenabviysxzdc9vj37kc9q4t...23MB/s]
> Downloads completed.
> ```

This is fixed now. It turns out that if you set output_dir with None it will use the current directory.

**Review by bgregoinductiva — APPROVED** [review]

**Review by sergiomsantos — APPROVED** [review]

**Pull request merged**